shell/platform/embedder/embedder_external_texture_gl_impeller.cc

```diff
@@ -0,0 +1,160 @@
+  if (!texture) {
+    return nullptr;
+  }
+  size_t width = size.width();
+  size_t height = size.height();
+  if (texture->width != 0 && texture->height != 0) {
+    width = texture->width;
+    height = texture->height;
+  }
+  if (texture->impeller_texture_type ==
+      FlutterGLImpellerTextureType::kFlutterGLImpellerTexturePixelBuffer) {
+    return ResolvePixelBufferTexture(texture.get(), context,
+                                     SkISize::Make(width, height));
+  } else {
+    return ResolveGpuSurfaceTexture(texture.get(), context,
+                                    SkISize::Make(width, height));
+  }
+}
+
+sk_sp<DlImage> EmbedderExternalTextureGLImpeller::ResolvePixelBufferTexture(
+    FlutterOpenGLTexture* texture,
+    PaintContext& context,
+    const SkISize& size) {
+  impeller::TextureDescriptor desc;
+  desc.type = impeller::TextureType::kTexture2D;
+  impeller::AiksContext* aiks_context = context.aiks_context;
+  const auto& gl_context =
+      impeller::ContextGLES::Cast(*aiks_context->GetContext());
+  desc.storage_mode = impeller::StorageMode::kDevicePrivate;
+  desc.format = impeller::PixelFormat::kR8G8B8A8UNormInt;
+  desc.size = {static_cast<int>(size.width()), static_cast<int>(size.height())};
+  desc.mip_count = 1;
+
+  auto textureGLES =
+      std::make_shared<impeller::TextureGLES>(gl_context.GetReactor(), desc);
+  if (!textureGLES->SetContents(texture->buffer, texture->buffer_size)) {
+    if (texture->destruction_callback) {
+      texture->destruction_callback(texture->user_data);
+    }
+    return nullptr;
+  }
+  if (texture->destruction_callback) {
+    texture->destruction_callback(texture->user_data);
+  }
+  return impeller::DlImageImpeller::Make(textureGLES);
+}
+
+sk_sp<DlImage> EmbedderExternalTextureGLImpeller::ResolveGpuSurfaceTexture(
+    FlutterOpenGLTexture* texture,
+    PaintContext& context,
+    const SkISize& size) {
+  impeller::TextureDescriptor desc;
+  desc.type = impeller::TextureType::kTextureExternalOES;
+  impeller::AiksContext* aiks_context = context.aiks_context;
+  const auto& gl_context =
+      impeller::ContextGLES::Cast(*aiks_context->GetContext());
+  desc.storage_mode = impeller::StorageMode::kDevicePrivate;
+  desc.format = impeller::PixelFormat::kR8G8B8A8UNormInt;
+  desc.size = {static_cast<int>(size.width()), static_cast<int>(size.height())};
+  desc.mip_count = 1;
+  auto textureGLES = std::make_shared<impeller::TextureGLES>(
+      gl_context.GetReactor(), desc,
+      impeller::TextureGLES::IsWrapped::kWrapped);
+  textureGLES->SetCoordinateSystem(
+      impeller::TextureCoordinateSystem::kUploadFromHost);
+  if (!textureGLES->Bind()) {
+    if (texture->destruction_callback) {
+      texture->destruction_callback(texture->user_data);
+    }
+    return nullptr;
+  }
+
+  if (!texture->bind_callback(texture->user_data)) {
+    if (texture->destruction_callback) {
+      texture->destruction_callback(texture->user_data);
+    }
+    return nullptr;
+  }
```

Comment:
1.
Here `destruction_callback` looks always nullptr. I don't know if there is a case where this callback is called.

2.
Can't we just use destruction_callback instead of bind_callback?
`BoolCallback bind_callback; ` I think we can remove `bind_callback`.
If OnBind() is guaranteed to be called after the call to PopulateTexture() in `bool ExternalTextureSurfaceEGLImpeller::OnBind()` in the embedder code, the bool return may not be necessary.